### PR TITLE
[server] WIP clean project parsing by using QgsProject

### PR DESCRIPTION
**Pull request by pblottiere**
The final aim of this PR is to remove project parsing done in server by using the QgsProject class instead. 

A `QgsServerProjectUtils` is introduced to retrieve specific entries from a project (using the same method than QgsLayerTreeUtils or QgsSymbolLayerUtils).

Tests are added at each stage to check the current behavior.

### CODE REVIEW

**Comment by rldhont** [comment]
@pblottiere do you think your work has to be merge before @dmarteau one ? https://github.com/qgis/QGIS/pull/3866

**Comment by pblottiere** [comment]
By doing tests around the `getCapabilities` request, I wondered:  
- for now, in case of WCS and WFS with no service url defined in the project, the WMS service url is used instead. Is it correct to do that?
- in the WMS server, the `SchemaLocation` entry indicates the WMS service url defined in the project, but not in WFS nor in WCS. Again, is it the correct behaviour?

**Comment by pblottiere** [comment]
@rldhont 

> do you think your work has to be merge before @dmarteau one ? #3866

As we said with @dmarteau in https://github.com/qgis/QGIS/pull/3866, it doesn't seem necessary.

File: python/server/qgsserverprojectutils.sip
Comment (elpaso) [inline python/server/qgsserverprojectutils.sip]: Add "from a QgsProject..."

File: src/server/qgsserver.h
Comment (elpaso) [inline src/server/qgsserver.h]: Unless it's something completely different in scope, I'd suggest to name this `mProjectRegistry` because `registry` is the word commonly used within the rest of QGIS for holding collections of objects.

File: src/server/qgsserverprojectparser.cpp
Comment (elpaso) [inline src/server/qgsserverprojectparser.cpp]: A general consideration here: I think it's ok to get the service URLs from the project, but they are not mandatory (and should remain optional), just make sure that the server can build and provide sensible default URLs for at least the core services.
@dmarteau maybe we should try to address this in the 3.0 refactoring: every service should be able to build its own URL (with a sensible default method in the base class).

**Review by elpaso — APPROVED** [review]
Good job! Thanks!

**Comment by dmarteau** [comment]
@pblottiere Just for your information: there is an extra QgsProject* parameter in the QgsService::executeRequest method, this parameter is not used atm but is the result of a discussion with mhugo for passing the project to the services when (wfs|wms|...)parsers will be removed.

**Comment by rldhont** [comment]
@pblottiere #4004 has been merged, so you have to update this PR.

File: src/server/qgsserverprojectparser.cpp
Comment (dmarteau) [inline src/server/qgsserverprojectparser.cpp]: Actually the build of service URL is done on a per-service basis from the requests url (no more getenv access). There is no exposed service url atm in QgsService interface but that can be easily added. 
See  https://github.com/qgis/QGIS/blob/master/src/server/services/wms/qgswmsutils.h#L56. This function is basically implemented for each services.

**Comment by dmarteau** [comment]
@pblottiere The heavy part of service refactoring as modules is over. This should culminate with this PR !

**Review by rldhont — CHANGES_REQUESTED** [review]
Rebase needed

**Comment by pblottiere** [comment]
@rldhont @dmarteau 

Actually I wanted to do more about cleaning the project parser but I was stuck with some QgsProject singleton in core. 

So I have to dig in to properly rebase (because some singleton has been removed since my last commits) and improve this PR. 

In other words, this PR is not ready to merge for now.

**Comment by rldhont** [comment]
@pblottiere what is your plan for this PR ? We can help if needed

**Comment by pblottiere** [comment]
@rldhont I was on another project since the beginning of the year but from today I have more time to work on Qgis and especially this PR. 

I let you know in the days to follow.

Thank you for your concern.

**Comment by wonder-sk** [comment]
Great work - looking forward to having it merged!

Few thoughts for this PR:
- QgsServerProjectUtils may be useful in qgis_core library - in QGIS desktop that may be useful for the server configuration GUIs
- QgsServerProjectUtils could take QgsProject in constructor, so that individual member functions do not need to ask for QgsProject
- for QgsMapLayer reading/writing I thought it may be useful to have something like QgsPathResolver object (instead of directly using QgsProject), the resolver would have just readPath/writePath methods. The idea is that map layer is not only read within the context of a project - it can be also read/written from/to QML files where we presumably want to resolve relative paths to the QML file's path rather than the current project... (not saying it has to be done within this PR, I can try to sort it out when this PR is merged)

**Review by sbrunner — APPROVED** [review]

**Review by dmarteau — COMMENTED** [review]
Wouldn't that be appropriate to move wmsServiceUrl local to the service ? Is there a specifiic reason to  get methods globally from  QgsServerProjectUtils
(same for wfs / wcs) ?

**Comment by pblottiere** [comment]
@dmarteau 

From my point of view, the aim of QgsServerProjectUtils is to contains everything concerning the project parsing.

**Review by rldhont — APPROVED** [review]
I think the @wonder-sk proposals about QgsPathResolver will be a great enhancement in QGIS environment.

**Comment by pblottiere** [comment]
@rldhont Please, do not merge this PR yet. I would like to check some thing first.

**Comment by rldhont** [comment]
@pblottiere I'm waiting for @dmarteau and your validation before merging

**Comment by dmarteau** [comment]
@wonder-sk I'm not even sure that QgsServerProjectUtils should be a class but only a collection of function in a namespace: I don't think we want removing parser and introduce a new kind of wrapper class around qgisproject. IMHO the functional approach makes more sens than OO.

**Comment by dmarteau** [comment]
@pblottiere  I understand: but this is is specifically related to services ?

**Comment by pblottiere** [comment]
@dmarteau 

>  I'm not even sure that QgsServerProjectUtils should be a class but only a collection of function in a namespace: I don't think we want removing parser and introduce a new kind of wrapper class around qgisproject. IMHO the functional approach makes more sens than OO in this case.

Indeed, why not.

>  I understand: but this is is specifically related to services ?
@wonder-sk was talking about moving it in core. But I don't have a lot of visibility on this point.

**Comment by dmarteau** [comment]
@pblottiere  please remove src/server/qgswmsserver.[cpp|h] and all others qgswfsserver/qgswcsserver: these files has been removed  since https://github.com/qgis/QGIS/pull/4019.

**Review by dmarteau — APPROVED** [review]

**Comment by rldhont** [comment]
@pblottiere I can merge when you want

**Comment by pblottiere** [comment]
@rldhont I'm moving the QgsServerProjectUtils as a namespace like @dmarteau advise me.

Then, it should be ok.

**Comment by pblottiere** [comment]
@dmarteau  I replaced the QgsServerProjectUtils class by a namespace. However, the build is failing on travis, due to the SERVER_EXPORT (but everything is compiling from my side, maybe a different version of gcc). 

And if I don't use the SERVER_EXPORT, functions within the namespace are not visible outside the qgis_server shared library, and so can't be used in wms/wcs/wfs modules nor in Python binding.

Do you have an idea? Or do we keep the class to avoid this?

**Comment by dmarteau** [comment]
@pblottiere you do not export a namespace: use SERVER_EXPORT on function declaration.

**Comment by pblottiere** [comment]
@dmarteau Thank you for your advice! Still plenty of things to learn...

**Comment by pblottiere** [comment]
@rldhont This PR is ready to merge. Sorry for the delay.

**Comment by dmarteau** [comment]
@pblottiere you are welcome :-) And thank you for your work.

**Comment by rldhont** [comment]
thanks @pblottiere

**Comment by sbrunner** [comment]
Many thanks @pblottiere :-)

File: src/server/services/DummyService/dummy.cpp
Comment (vmora) [inline src/server/services/DummyService/dummy.cpp]: What is the meaning of a null ptr there ? If none, why not a const & rather than a const (naked) ptr ?

File: src/server/services/wcs/qgswcsgetcapabilities.h
Comment (vmora) [inline src/server/services/wcs/qgswcsgetcapabilities.h]: * .vs. & -> to me ptr (smart to indicate who is resp. for the resource) indicate the possibility for null, I'd use & everywhere else. Is the function able to do something meaningfull without the serverIface... maybe, without a project... maybe. If its the case, doc, if it's not &

**Comment by dmarteau** [comment]
@vmora: because of history: when this function was added in https://github.com/qgis/QGIS/pull/3866. QgsProject was still a singleton and thus this parameters was only there to make room for passing QgsProject instance when it well be ready and using reference was not an option then.

**Pull request merged**
